pkg/collector/corechecks/ebpf/c/tcp-queue-length-kern.c

```diff
@@ -0,0 +1,145 @@
+#include <linux/compiler.h>
+
+#include <linux/kconfig.h>
+#include <linux/ptrace.h>
+#include <linux/types.h>
+#include <linux/version.h>
+#include <linux/tcp.h>
+
+#define bpf_printk(fmt, ...)                       \
+	({                                             \
+		char ____fmt[] = fmt;                      \
+		bpf_trace_printk(____fmt, sizeof(____fmt), \
+						 ##__VA_ARGS__);           \
+	})
+
+#include "bpf_helpers.h"
+#include "bpf-common.h"
+#include "tcp-queue-length-kern-user.h"
+
+/*
+ * The `tcp_queue_stats` map is used to share with the userland program system-probe
+ * the statistics (max size of receive/send buffer)
+ */
+
+struct bpf_map_def SEC("maps/tcp_queue_stats") tcp_queue_stats = {
+    .type = BPF_MAP_TYPE_PERCPU_HASH,
+    .key_size = sizeof(struct stats_key),
+    .value_size = sizeof(struct stats_value),
+    .max_entries = 1024,
+    .pinning = 0,
+    .namespace = "",
+};
+
+/*
+ * the `who_recvmsg` and `who_sendmsg` maps are used to remind the sock pointer
+ * received as input parameter when we are in the kretprobe of tcp_recvmsg and tcp_sendmsg.
+ */
+struct bpf_map_def SEC("maps/who_recvmsg") who_recvmsg = {
+    .type = BPF_MAP_TYPE_HASH,
+    .key_size = sizeof(u64),
+    .value_size = sizeof(struct sock*),
+    .max_entries = 100,
+    .pinning = 0,
+    .namespace = "",
+};
+
+struct bpf_map_def SEC("maps/who_sendmsg") who_sendmsg = {
+    .type = BPF_MAP_TYPE_HASH,
+    .key_size = sizeof(u64),
+    .value_size = sizeof(struct sock*),
+    .max_entries = 100,
+    .pinning = 0,
+    .namespace = "",
+};
+
+// TODO: replace all `bpf_probe_read` by `bpf_probe_read_kernel` once we can assume that we have at least kernel 5.5
+static __always_inline int check_sock(struct sock* sk) {
+    struct stats_value zero = {
+        .read_buffer_max_usage = 0,
+        .write_buffer_max_usage = 0
+    };
+
+    struct stats_key k;
+    get_cgroup_name(k.cgroup_name, sizeof(k.cgroup_name));
+
+    struct stats_value* v = bpf_map_lookup_elem(&tcp_queue_stats, &k);
+    if (v == NULL) {
```

Comment:
See other comment about ordering of `bpf_map_lookup_elem` and `bpf_map_update_elem`.